File: CS/ViewModels/DragDropViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Mvvm.POCO;
using System.Data;
using System.Windows.Forms;
using System.ComponentModel;
using DragDropTwoGrids.Model;

namespace DragDropTwoGrids.ViewModels {
    public class DragDropViewModel {
        public virtual BindingList<File> Files { get; set; }
        public DragDropViewModel() {
            InitData();
        }
        private void InitData() {
            Files = Repository.CreateFiles();
        }
        public void Drop(MyDragAndDropEventArgs dataObject) {
            File file = dataObject.Record as File;
            if(file != null) {
                if(Files.Where(f => f.FileName == file.FileName && f.FileSize == file.FileSize).FirstOrDefault() != null) {
                    dataObject.Cancel = true;
                }
                else {
                    dataObject.Cancel = false;
                    Files.Add(file);
                }
            }
        }
        public void RemoveRecord(object record) {
            File file = record as File;
            if(file != null)
                Files.Remove(file);
        }
    }
}

File: VB/ViewModels/DragDropViewModel.vb
Imports System.Linq
Imports System.ComponentModel
Imports DragDropTwoGrids.Model

Namespace DragDropTwoGrids.ViewModels

    Public Class DragDropViewModel

        Public Overridable Property Files As BindingList(Of File)

        Public Sub New()
            InitData()
        End Sub

        Private Sub InitData()
            Files = CreateFiles()
        End Sub

        Public Sub Drop(ByVal dataObject As MyDragAndDropEventArgs)
            Dim file As File = TryCast(dataObject.Record, File)
            If file IsNot Nothing Then
                If Files.Where(Function(f) Equals(f.FileName, file.FileName) AndAlso f.FileSize = file.FileSize).FirstOrDefault() IsNot Nothing Then
                    dataObject.Cancel = True
                Else
                    dataObject.Cancel = False
                    Files.Add(file)
                End If
            End If
        End Sub

        Public Sub RemoveRecord(ByVal record As Object)
            Dim file As File = TryCast(record, File)
            If file IsNot Nothing Then Files.Remove(file)
        End Sub
    End Class
End Namespace
